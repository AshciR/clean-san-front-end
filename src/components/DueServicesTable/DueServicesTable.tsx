import { Chip, Table, TableBody, TableCell, TableContainer, TableHead, TableRow, Typography } from '@mui/material';
import { DateTime } from 'luxon';
import { FC } from 'react';
import DueService from '../../shared/DueService.model';
import ServiceStatus from '../../shared/ServiceStatus.model';
import styles from './DueServicesTable.module.scss';

interface DueServicesTableProps {
  dueServices: Array<DueService>
};

const DueServicesTable: FC<DueServicesTableProps> = ({ dueServices }: DueServicesTableProps) => {

  const hasServices = dueServices.length !== 0;

  return (
    <div className={styles.DueServicesTable} data-testid="due-services-table">
      {hasServices ? <VisableDueServiceTable services={dueServices} /> : <NoDueServicesDisplay />}
    </div>
  );
};

interface VisableDueServiceTableProps {
  services: Array<DueService>
};

const VisableDueServiceTable = ({ services }: VisableDueServiceTableProps) => (
  <TableContainer data-testid="visable-due-services-table">
    <Table aria-label="due services table">
      <TableHead>
        <TableRow>
          <TableCell>Id</TableCell>
          <TableCell align='right'>Client</TableCell>
          <TableCell align='right'>Frequency</TableCell>
          <TableCell align='right'>Due Date</TableCell>
          <TableCell align='right'>Status</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        {services.map(service => <DueServiceRow key={service.id} service={service} />)}
      </TableBody>
    </Table>
  </TableContainer>
);

interface DueServiceRowProps {
  service: DueService
};

const DueServiceRow = ({ service }: DueServiceRowProps) => (
  <TableRow data-testid="due-service-table-row">
    <TableCell>{service.id}</TableCell>
    <TableCell align='right'>{service.client.name}</TableCell>
    <TableCell align='right'>{service.contract.serviceFrequency}</TableCell>
    <TableCell align='right'>{service.dueDate.toLocaleString(DateTime.DATE_MED)}</TableCell>
    <TableCell align='right'>{getStatusComponent(service.currentStatus)}</TableCell>
  </TableRow>
)

const NoDueServicesDisplay = () => (
  <Typography variant='h5' data-testid="no-due-services-display">There are no due services at this time</Typography>
);

const getStatusComponent = (currentStatus: ServiceStatus) => {
  switch (currentStatus) {
    case ServiceStatus.NOT_COMPLETED:
      return <NotCompletedServiceStatus />;
    case ServiceStatus.IN_PROGRESS:
      return <InProgressServiceStatus />;
    case ServiceStatus.COMPLETED:
      return <CompletedServiceStatus />
    case ServiceStatus.CANCELLED:
      return <CancelledServiceStatus />
    default:
      new Error(`No status indicator for ${currentStatus} exists`);
  }
};

const NotCompletedServiceStatus = () => (
  <Chip label='Not Completed' color='warning' variant="outlined" size='small'></Chip>
);

const InProgressServiceStatus = () => (
  <Chip label='In Progress' color='info' variant="outlined" size='small'></Chip>
);

const CompletedServiceStatus = () => (
  <Chip label='Completed' color='success' variant="outlined" size='small'></Chip>
);

const CancelledServiceStatus = () => (
  <Chip label='Cancelled' variant="outlined" size='small'></Chip>
);

export default DueServicesTable;


