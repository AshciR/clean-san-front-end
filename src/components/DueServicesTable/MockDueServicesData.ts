import DueService, { createDefaultDueService } from "../../shared/DueService.model";

const service1: DueService = createDefaultDueService();
const service2: DueService = {
    ...service1,
    id: service1.id + 1,
    dueDate: service1.dueDate.minus({ months: 1 }),
};
const service3: DueService = {
    ...service2,
    id: service2.id + 1,
    dueDate: service2.dueDate.minus({ months: 1 }),
};

const MOCK_DUE_SERVICES: Array<DueService> = [
    service1, service2, service3
];

export default MOCK_DUE_SERVICES;