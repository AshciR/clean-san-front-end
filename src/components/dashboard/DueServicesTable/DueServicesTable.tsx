import {
  Box,
  Button,
  FormControl,
  MenuItem,
  Select,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TablePagination,
  TableRow,
  TableSortLabel,
  Typography
} from '@mui/material';
import {DateTime} from 'luxon';
import {FC} from 'react';
import DueService from '../../../shared/DueService.model';
import ServiceStatus from '../../../shared/ServiceStatus.model';
import styles from './DueServicesTable.module.scss';
import ServiceStatusChip, {STATUS_CHIP_WIDTH} from "../ServiceStatusChip/ServiceStatusChip";
import ServiceFrequencyChip from "../../shared/ServiceFrequencyChip/ServiceFrequencyChip";
import {DashboardOrderByOptions, ITEMS_PER_PAGE_OPTIONS, SortOrder} from "../DashboardPage/dashboardPage.reducer";

const SERVICE_ID_WIDTH = 100;
const STATUS_COLUMN_WIDTH = STATUS_CHIP_WIDTH + 60;

interface DueServicesTableProps {
  dueServices: Array<DueService>
  totalServices: number
  servicesPerPage: number
  currentPage: number
  sortOrder: SortOrder
  handleUpdateService: (updatedService: DueService) => void
  handleOpenViewAssociatedServicesModal: (selectedService: DueService) => void
  handleChangePage: (newPageNumber: number) => void
  handleChangeRowsPerPage: (newRowsPerPage: number) => void
  handleSortBy: (sortColumn: string) => void
}

const DueServicesTable: FC<DueServicesTableProps> = ({
                                                       dueServices,
                                                       totalServices,
                                                       servicesPerPage,
                                                       currentPage,
                                                       sortOrder,
                                                       handleUpdateService,
                                                       handleOpenViewAssociatedServicesModal,
                                                       handleChangePage,
                                                       handleChangeRowsPerPage,
                                                       handleSortBy
                                                     }: DueServicesTableProps) => {

  const hasServices = dueServices.length !== 0;

  return (
    <div className={styles.DueServicesTable} data-testid="due-services-table">
      {hasServices ? <VisibleDueServiceTable
          services={dueServices}
          totalServices={totalServices}
          servicesPerPage={servicesPerPage}
          currentPage={currentPage}
          sortOrder={sortOrder}
          handleUpdateService={handleUpdateService}
          handleOpenViewAssociatedServicesModal={handleOpenViewAssociatedServicesModal}
          handleChangePage={handleChangePage}
          handleChangeRowsPerPage={handleChangeRowsPerPage}
          handleSortBy={handleSortBy}
        /> :
        <NoDueServicesDisplay/>}
    </div>
  );
};

interface VisibleDueServiceTableProps {
  services: Array<DueService>
  totalServices: number
  servicesPerPage: number
  currentPage: number
  sortOrder: SortOrder
  handleUpdateService: (updatedService: DueService) => void
  handleOpenViewAssociatedServicesModal: (selectedService: DueService) => void
  handleChangePage: (newPageNumber: number) => void
  handleChangeRowsPerPage: (newRowsPerPage: number) => void
  handleSortBy: (sortColumn: string) => void
}

const VisibleDueServiceTable = ({
                                  services,
                                  totalServices,
                                  servicesPerPage,
                                  currentPage,
                                  sortOrder,
                                  handleUpdateService,
                                  handleOpenViewAssociatedServicesModal,
                                  handleChangePage,
                                  handleChangeRowsPerPage,
                                  handleSortBy
                                }: VisibleDueServiceTableProps) => (

  <Box>
    <TableContainer
      data-testid="visible-due-services-table"
      sx={{maxHeight: 550}}
    >
      <Table
        aria-label="due services table"
        stickyHeader={true}
      >
        <TableHead>
          <TableRow>
            <TableCell sx={{width: SERVICE_ID_WIDTH}}>Service Id</TableCell>
            <TableCell>
              <TableSortLabel
                active={sortOrder.orderBy === DashboardOrderByOptions.CLIENT}
                direction={sortOrder.direction}
                onClick={() => handleSortBy(DashboardOrderByOptions.CLIENT)}
              />
              Client
            </TableCell>
            <TableCell>Frequency</TableCell>
            <TableCell>Due Date</TableCell>
            <TableCell sx={{width: STATUS_COLUMN_WIDTH}}>Current Status</TableCell>
            <TableCell sx={{width: STATUS_COLUMN_WIDTH}}>New Status</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {services.map(service =>
            <DueServiceRow
              key={service.id}
              service={service}
              handleUpdateService={handleUpdateService}
              handleOpenViewAssociatedServicesModal={handleOpenViewAssociatedServicesModal}
            />
          )}
        </TableBody>
      </Table>
    </TableContainer>
    <TablePagination
      rowsPerPageOptions={ITEMS_PER_PAGE_OPTIONS}
      component="div"
      count={totalServices}
      rowsPerPage={servicesPerPage}
      page={currentPage}
      onPageChange={(event, newPage) => {
        handleChangePage(newPage);
      }}
      onRowsPerPageChange={(event) => {
        const newRowsPerPage = parseInt(event.target.value, 10);
        handleChangeRowsPerPage(newRowsPerPage)
      }}
    />
  </Box>
);

interface DueServiceRowProps {
  service: DueService
  handleUpdateService: (updatedService: DueService) => void
  handleOpenViewAssociatedServicesModal: (selectedService: DueService) => void
}

const DueServiceRow = ({service, handleUpdateService, handleOpenViewAssociatedServicesModal}: DueServiceRowProps) => (
  <TableRow data-testid="due-service-table-row">
    <TableCell sx={{width: SERVICE_ID_WIDTH}}>
      <Button
        size="small"
        color="primary"
        onClick={() => handleOpenViewAssociatedServicesModal(service)}
      >
        {service.id}
      </Button>
    </TableCell>
    <TableCell>{service.client.name}</TableCell>
    <TableCell>
      <ServiceFrequencyChip frequency={service.contract.serviceFrequency}/>
    </TableCell>
    <TableCell>{service.dueDate.toLocaleString(DateTime.DATE_MED)}</TableCell>
    <TableCell sx={{width: STATUS_COLUMN_WIDTH}}>
      <ServiceStatusChip status={service.currentStatus}/>
    </TableCell>
    <TableCell sx={{width: STATUS_COLUMN_WIDTH}}>
      <StatusDropDown service={service} handleUpdateService={handleUpdateService}/>
    </TableCell>
  </TableRow>
)

const NoDueServicesDisplay = () => (
  <Typography variant='h5' data-testid="no-due-services-display">There are no due services at this time</Typography>
);


interface StatusDropDownProps {
  service: DueService
  handleUpdateService: (updatedService: DueService) => void
}

const StatusDropDown = ({service, handleUpdateService}: StatusDropDownProps) => {
  return (
    <Box>
      <FormControl sx={{width: STATUS_COLUMN_WIDTH}}>
        <Select
          id={`new-status-select-${service.id}`}
          value={service?.prospectiveStatus || service.currentStatus}
          onChange={(event) => {
            const updatedService = {
              ...service,
              prospectiveStatus: event.target.value as ServiceStatus
            };
            handleUpdateService(updatedService)

          }}
          variant='outlined'
          defaultValue={service.currentStatus}
          sx={{
            '& .MuiSelect-select': {paddingTop: 1, paddingBottom: 1} // 8px
          }}
        >
          {
            Object.keys(ServiceStatus).map((status) => (
              <MenuItem key={status} value={status}>
                <ServiceStatusChip status={status as ServiceStatus}/>
              </MenuItem>
            ))
          }
        </Select>
      </FormControl>
    </Box>
  );
};

export default DueServicesTable;


