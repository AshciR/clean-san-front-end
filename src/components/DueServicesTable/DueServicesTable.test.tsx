import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import DueServicesTable from './DueServicesTable';
import MOCK_DUE_SERVICES from './MockDueServicesData';
import { DateTime } from 'luxon';
import DueService from '../../shared/DueService.model';


describe('<DueServicesTable />', () => {

  it('it should mount', () => {
    render(<DueServicesTable dueServices={[]} handleUpdateService={(updatedService: DueService) => {}}/>);

    const dueServicesTable = screen.getByTestId('due-services-table');

    expect(dueServicesTable).toBeInTheDocument();
  });

  it('it should display due services', () => {

    // When: the Due Services Table is rendered with services 
    render(<DueServicesTable dueServices={MOCK_DUE_SERVICES} handleUpdateService={(updatedService: DueService) => {}} />);

    // Then: All the rows should have the correct info
    const dueServicesTableRows = screen.getAllByTestId('due-service-table-row');
    expect(dueServicesTableRows.length).toBe(MOCK_DUE_SERVICES.length);
    dueServicesTableRows.forEach((row, index) => assertRowContent(row, MOCK_DUE_SERVICES[index]));

  });

  it('it should display message when there are no due services', () => {

    // When: the Due Services Table is rendered without any services 
    render(<DueServicesTable dueServices={[]} handleUpdateService={(updatedService: DueService) => {}} />);

    // Then: The no services display message should be present
    const noDueServicesDisplay = screen.getByText('There are no due services at this time');
    expect(noDueServicesDisplay).toBeInTheDocument();

  });

  // TODO: Add test to check that handler was called

  const assertRowContent = (row: HTMLElement, service: DueService) => {

    const idColumn = row.getElementsByTagName('td')[0];
    expect(idColumn.textContent).toBe(service.id.toString());

    const clientColumn = row.getElementsByTagName('td')[1];
    expect(clientColumn.textContent).toBe(service.client.name);

    const frequencyColumn = row.getElementsByTagName('td')[2];
    expect(frequencyColumn.textContent).toBe(service.contract.serviceFrequency);

    const dueDateColumn = row.getElementsByTagName('td')[3];
    expect(dueDateColumn.textContent).toBe(service.dueDate.toLocaleString(DateTime.DATE_MED));

    const statusColumn = row.getElementsByTagName('td')[4];
    expect(statusColumn.textContent?.toLocaleLowerCase()).toBe(service.currentStatus.toLocaleLowerCase());

  }

});