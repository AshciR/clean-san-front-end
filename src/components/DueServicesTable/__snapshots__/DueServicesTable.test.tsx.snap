// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DueServicesTable /> Status Chip component shows correctly 1`] = `
<div
  class="DueServicesTable"
  data-testid="due-services-table"
>
  <div
    class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
    data-testid="visable-due-services-table"
  >
    <table
      aria-label="due services table"
      class="MuiTable-root css-rqglhn-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Id
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
          >
            Client
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
          >
            Frequency
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
          >
            Due Date
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
          >
            Status
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          data-testid="due-service-table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            100
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            Rick & Morty Adventures
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            MONTHLY
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            Mar 27, 2022
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            <div
              class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorWarning MuiChip-outlinedWarning css-mpai30-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
              >
                Not Completed
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          data-testid="due-service-table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            101
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            Rick & Morty Adventures
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            MONTHLY
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            Feb 27, 2022
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            <div
              class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorInfo MuiChip-outlinedInfo css-1wvzl5v-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
              >
                In Progress
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          data-testid="due-service-table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            102
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            Rick & Morty Adventures
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            MONTHLY
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            Jan 27, 2022
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            <div
              class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorSuccess MuiChip-outlinedSuccess css-1smg64b-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
              >
                Completed
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          data-testid="due-service-table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            103
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            Rick & Morty Adventures
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            MONTHLY
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            Dec 27, 2021
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            <div
              class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-911tg4-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
              >
                Cancelled
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
